### ISSUE TITLE
Wrong lint message when referencing from a wrong location

### ISSUE THREAD

**Issue by tatomyr** [Redocly/redocly-cli#767]
**Describe the bug**

If a valid part of a definition is being referenced from a section that does not expect this syntax, the linter shows an incorrect message inside the valid part.

**To Reproduce**

Consider this definition: 

```
# openapi.yaml

openapi: 3.0.0
components:
  requestBodies:
    CreateContainer:
      content:
        application/json:
          schema:
            type: object
  schemas:
    PipeContainerCreate:
      title: PipelineContainerCreate
      allOf:
      - "$ref": "#/components/requestBodies/CreateContainer"
   ```

Run `redocly lint openapi.yaml  --extends minimal`

Although the `requestBody` could receive the `content` field, linter notes that `Property `content` is not expected here.`.

**Expected behavior**

We need to handle this common issue and show a proper message and a "referenced from".

**Redocly Version(s)**
`1.0.0-beta.104`

**`Node.js` Version(s)**
`v16.14.0`

**Additional context**
See this [conversation](https://redoc-ly.slack.com/archives/C02UHE7EL3E/p1657730049335399) for additional context.

**Comment by AntonKozachuk** [comment]
@adamaltman @RomanHotsiy, would use your opinion here.

I was able to add an option to show the `referenced from` part for the spec rule in case we are reporting on the `$ref`.

Example:
<img width="881" alt="image" src="https://user-images.githubusercontent.com/54616703/186248494-0fa54ebe-754e-4548-a6e5-a713987384de.png">

However, I have a few things to consider:
1) We won't be able to improve the message right in the `spec` rule (or I didn't find a way yet), I think it might be possible to add it with another rule (I'm still investigating). Do you think we should continue working on this and notify users about the wrong `$ref` in Schema?
2) Should we show this `referenced from` part on every report in the `spec` rule? Or investigate adding this to other rules too?

**Comment by RomanHotsiy** [comment]
Can you open a PR for `referenced from` part.This would be useful anyways. I would add it to all messages.

Then I would better focus on another tasks. This one is taking too much time.

**Comment by adamaltman** [comment]
I like what @RomanHotsiy proposes. That would be very helpful as you've shown in the screenshot.

**Comment by benatdelta** [comment]
Fixes [#843]
